piff/knn_interp.py

```diff
@@ -0,0 +1,373 @@
+            # if no y, then interpolate
+            y = numpy.array([self.knn[key].predict(X) for key in self.attr_target]).T
+        if logger:
+            logger.debug('Regression shape: %s', y.shape)
+        # add misalignment shape (n_targets, 3)
+        # X is (n_samples, 2)
+        # y is (n_samples, n_targets)
+        y = y + self.misalignment[numpy.newaxis, :, 0] \
+                + X[:, 1, numpy.newaxis] * self.misalignment[numpy.newaxis, :, 1] \
+                + X[:, 0, numpy.newaxis] * self.misalignment[numpy.newaxis, :, 2]
+
+        return y
+
+    def writeSolution(self, fits, extname):
+        """Write the solution to a FITS binary table.
+
+        Save the knn params and the X and y arrays
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
+
+        dtypes = [('X', self.X.dtype, self.X.shape), ('Y', self.y.dtype, self.y.shape),
+                  ('XPIXEL', self.Xpixel.dtype, self.Xpixel.shape),
+                  ('ATTR_TARGET', self.attr_target.dtype, self.attr_target.shape),
+                  ('ATTR_INTERP', self.attr_interp.dtype, self.attr_interp.shape),
+                  ('MISALIGNMENT', self.misalignment.dtype, self.misalignment.shape)]
+        data = numpy.empty(1, dtype=dtypes)
+        # assign
+        data['X'] = self.X
+        data['Y'] = self.y
+        data['XPIXEL'] = self.Xpixel
+        data['ATTR_TARGET'] = self.attr_target
+        data['ATTR_INTERP'] = self.attr_interp
+        data['MISALIGNMENT'] = self.misalignment
+
+        # put the knn params in the header?
+        header = self.kwargs
+
+        # write to fits
+        fits.write_table(data, extname=extname, header=header)
+
+    def readSolution(self, fits, extname):
+        """Read the solution from a FITS binary table.
+
+        The extension should contain the same values as are saved
+        in the writeSolution method.
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
+        # header = fits[extname].read_header()
+        data = fits[extname].read()
+
+        # kwargs come from base class read()
+        # self.kwargs = header
+
+        # attr_target and attr_interp assigned in build
+        self.build(data['ATTR_INTERP'][0], data['ATTR_TARGET'][0])
+        # self.X and self.y assigned in _fit
+        self._fit(data['X'][0], data['Y'][0])
+        self.misalign_wavefront(data['MISALIGNMENT'][0])
+
+        # other attributes
+        self.Xpixel = data['XPIXEL'][0]
+
+
+# pretty sure the correct way to do this is in the WCS framework, but let's
+# move forward first:
+def pixel_to_focal(stardata):
+    """Take stardata and add focal plane position to properties
+
+    :param stardata:    The stardata with property 'ccdnum'
+
+    :returns stardata:  New stardata with updated properties
+    """
```

Comment:
I don't know if 10-15 arcsec is good enough, but the WCS solutions are going to be right to < 1 arcsec for sure (Usually < 0.02 arcsec in fact).  So if the telescope thinks it's pointing at some other location than it really is, I think that means (u,v) = (0,0) will be at a different point on the focal plane.

However, perhaps this really means that we should have an alternate way of specifying the (u,v) = (0,0) position for des exposures.  Like use the sky coordinate corresponding to (-20,2048) on chip 35.  (Or something like this -- I'm just eyeballing it, but you probably have a better estimate that we could use.)

Then in the input field of the config we can have a few options for how to specify the pointing rather than just the ra,dec columns of the fits header (of the first file in the list...).  Like
```
input:
    ...
    pointing:
        type: DECam
```